drivers/firmware/tisci/tisci.c

```diff
@@ -0,0 +1,1599 @@
+{
+	struct tisci_data *data = dev->data;
+	struct tisci_xfer *xfer = &data->xfer;
+	struct tisci_msg_hdr *hdr;
+
+	if (rx_message_size > data->desc.max_msg_size ||
+	    tx_message_size > data->desc.max_msg_size ||
+	    (rx_message_size > 0 && rx_message_size < sizeof(*hdr)) ||
+	    tx_message_size < sizeof(*hdr)) {
+		return NULL;
+	}
+
+	k_sem_take(&data->data_sem, K_FOREVER);
+	data->seq++;
+	k_sem_give(&data->data_sem);
+
+	xfer->tx_message.data = req_buf;
+	xfer->tx_message.size = tx_message_size;
+	xfer->rx_message.buf = resp_buf;
+	xfer->rx_message.size = (uint8_t)rx_message_size;
+
+	hdr = (struct tisci_msg_hdr *)req_buf;
+	hdr->seq = data->seq;
+	hdr->type = msg_type;
+	hdr->host = data->host_id;
+	hdr->flags = msg_flags;
+
+	if (rx_message_size) {
+		hdr->flags = hdr->flags | TISCI_FLAG_REQ_ACK_ON_PROCESSED;
+	}
+
+	return xfer;
+}
+
+static void callback(const struct device *dev, mbox_channel_id_t channel_id, void *user_data,
+		     struct mbox_msg *data)
+{
+	k_sem_give(&response_ready_sem);
+}
+
+static bool tisci_is_response_ack(void *r)
+{
+	struct tisci_msg_hdr *hdr = (struct tisci_msg_hdr *)r;
+
+	return hdr->flags & TISCI_FLAG_RESP_GENERIC_ACK ? true : false;
+}
+
+static int tisci_get_response(const struct device *dev, struct tisci_xfer *xfer)
+{
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	struct tisci_data *dev_data = dev->data;
+	struct tisci_msg_hdr *hdr;
+
+	if (!xfer->rx_message.buf) {
+		LOG_ERR("No response buffer provided");
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&response_ready_sem, K_MSEC(dev_data->desc.max_rx_timeout_ms)) != 0) {
+		LOG_ERR("Timeout waiting for response");
+		return -ETIMEDOUT;
+	}
+
+	if (xfer->rx_message.size > dev_data->desc.max_msg_size) {
+		LOG_ERR("rx_message.size [ %d ] > max_msg_size\n", xfer->rx_message.size);
+		return -EINVAL;
+	}
+
+	if (rx_message.size < xfer->rx_message.size) {
+		LOG_ERR("rx_message.size [ %d ] < xfer->rx_message.size\n", rx_message.size);
+		return -EINVAL;
+	}
+
+	memcpy(xfer->rx_message.buf, rx_message.buf, xfer->rx_message.size);
+	hdr = (struct tisci_msg_hdr *)xfer->rx_message.buf;
+
+	/* Sanity check for message response */
+	if (hdr->seq != dev_data->seq) {
+		LOG_ERR("HDR seq != data seq [%d != %d]\n", hdr->seq, dev_data->seq);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int tisci_do_xfer(const struct device *dev, struct tisci_xfer *xfer)
+{
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	const struct tisci_config *config = dev->config;
+	struct mbox_msg *msg = &xfer->tx_message;
+	int ret;
+
+	ret = mbox_send_dt(&config->mbox_tx, msg);
+	if (ret < 0) {
+		LOG_ERR("Could not send (%d)\n", ret);
+		return ret;
+	}
+
+	/* Get response if requested */
+	if (xfer->rx_message.size) {
+		ret = tisci_get_response(dev, xfer);
+		if (ret) {
+			return ret;
+		}
+		if (!tisci_is_response_ack(xfer->rx_message.buf)) {
+			LOG_ERR("TISCI Response in NACK\n");
+			return -ENODEV;
+		}
+	}
+
+	return 0;
+}
+
+static int tisci_init(const struct device *dev)
```

Comment:
minor suggestion: can move `tisci_init` to the end of file, just above device init